---
title: "Examining the draft success of NFL teams using Pro Football Reference's Approximate Value statistic"
author: "Myron Keith Gibert Jr"
date: "March 4, 2020"
output: pdf_document
toc: true
---

## Introduction
For this first programming assignment I wrote three functions that are meant to interact with dataset that accompanies this assignment. The dataset is contained in a zip file specdata.zip that is included in the GitHub repository.

## Set parameters

```{r parameters}
countfilename <- "TUCR_counts_80.txt"

makekpmplots <- TRUE

makeRPKMboxplots <- TRUE

repel <- TRUE

outputdir <- "Outputs_80"

useintermediate <- TRUE

```

## Install required programs

```{r setup, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)

if (!require("tidyverse")) install.packages("tidyverse")
library("tidyverse")

if (!require("ggplot2")) install.packages("ggplot2")
library("ggplot2")

if (!require("ggforce")) install.packages("ggforce")
library("ggforce")

if (!require("ggrepel")) install.packages("ggrepel")
library("ggrepel")

if (!require("ggfortify")) install.packages("ggfortify")
library("ggfortify")

if (!require("ggdendro")) install.packages("ggdendro")
library("ggdendro")

if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if (!requireNamespace("DESeq2", quietly = TRUE))BiocManager::install("DESeq2")
library("DESeq2")

if (!require("survival")) install.packages("survival")
library("survival")

if (!require("broom")) install.packages("broom")
library("broom")

if (!requireNamespace("limma", quietly = TRUE))BiocManager::install("limma")
library("limma")

if (!require("Tmisc")) install.packages("Tmisc")
library("Tmisc")

if (!require("survminer")) install.packages("survminer")
library("survminer")

if (!require("corrr")) install.packages("corrr")
library("corrr")

if (!require("here")) install.packages("here")
library("here")

if (!require("lessR")) install.packages("lessR")
library("lessR")

```

```{r outputs}
if(!dir.exists(outputdir)){
  dir.create(outputdir)
}
```


```{r tinytex, include=FALSE}
## INSTALLING LATEX FOR RMARKDOWN

#RMarkdown requires LaTex to create pdf documents as an output. More information can be found [here](https://bookdown.org/yihui/rmarkdown/pdf-document.html). Alternatively, output can be set to "html_document" or "word_document". End users may install LaTex by setting the "wanttinytex" variable to FALSE and running the following script:

#Install tinytex to let RMarkdown create a pdf document? Default: wanttinytex <- FALSE
wanttinytex <- TRUE

if(tinytex:::is_tinytex()==FALSE && wanttinytex == TRUE){
if (!require("tinytex")) install.packages("tinytex")
tinytex::install_tinytex()
}
```

## Stringtie

```{bash stringtie, eval = FALSE}
wget ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz
mv Homo_sapiens.GRCh38.99.gtf.gz stringtie
gunzip stringtie/Homo_sapiens.GRCh38.99.gtf.gz

module load gcc/7.1.0
module load stringtie/2.0.6

for bam in $(find GBM-RNASEQ-RAW/* -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
name2=$(basename "$name")
if [ ! -f stringtie/$name2.gtf ] 
then
    echo "$name2.gtf does not exist"
    stringtie $bam -o $name.gtf
    mv $name.gtf ./stringtie
else
    echo "$name2.gtf exists... skipping"
fi

if [ ! -f $name.gtf ]; then
  stringtie $bam -o $name.gtf
fi
mv $name.gtf ./stringtie
done

cd stringtie

stringtie --merge *rehead.gtf -G Homo_sapiens.GRCh38.99.gtf -o stringtie_merged.gtf

intersectBed -a hg38.ultraConserved.bed -b CHESSgenes.bed -wa -wb > intersectchess_TUCRs.bed

intersectBed -a hg38.ultraConserved.bed -b stringtie_merged.gtf -wa -wb > intersectstringtie_TUCRs.bed

cp stringtie_merged.gtf stringtie_merged.bed
```

## Setup for TCGA RNA-Seq analysis

```{bash tcga, eval = FALSE}
## Create directory for all outputs from TCGA analysis data
mkdir tcga_analysis/
```

## Extract headers from BAM files

```{bash tcgaheader, eval = FALSE}
## Extract header information from BAM files
# Create directory for extracted headerfiles
mkdir tcga_analysis/head/

# Create extracted header files
for bam in $(find GBM-RNASEQ-RAW/* -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools view -H $bam > $name.headerids.txt
mv $name.headerids.txt ./tcga_analysis/head/
done
```

## Extract TCGA patient barcodes from BAM headers

```{bash tcgaids, eval = FALSE}
## Get new file names for BAM files using TCGA IDs from extracted header
mkdir tcga_analysis/ids/

# Initialize summary file with TCGA IDs
echo "RAWID,TCGAID" > tcga_analysis/ids/tcgaIDs.csv 

# Extract IDs from header files
for header in $(find tcga_analysis/head/* -name '*headerids.txt')
do
name=$(echo $header | awk -F ".headerids.txt" '{print $1}')
name2=$(echo $name | cut -c20-)
ids1=$(awk '$1=="@RG"{print $3}' $header)
ids2=$(echo $ids1 | cut -c4-)
echo "$name2.bam,$ids2.bam"
echo "$name2.bam,$ids2.bam" > $name.tcgaid.csv
echo "$name2.bam,$ids2.bam" >> tcga_analysis/ids/tcgaIDs.csv 
mv $name.tcgaid.csv ./tcga_analysis/ids/
done
```

## Reindex TCGA BAM files under new names

```{bash tcgaindex, eval = FALSE}

##Rename BAM files using TCGA ids
# Create folder for new files
mkdir tcga_analysis/reindex/
# Copy BAM files into new directory with updated names

while IFS=, read orig target; do
orig2=$(find GBM-RNASEQ-RAW/* -name $orig)
echo $orig2
cp $orig2 $target
mv $target tcga_analysis/reindex/
done < tcga_analysis/ids/tcgaIDs.csv

# Reindex the renamed files using samtools.
for bam in $(find tcga_analysis/reindex/ -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools index $bam
done
```

## Generating count tables for survival and differential expression.

```{bash tcgacounts, eval = FALSE}
mkdir tcga_analysis/foldchange

#Write header line
echo -e chrom"\t"start"\t"end"\t"id tcga_analysis/reindex/*bam | sed -e "s/ /\t/g" > tcga_analysis/foldchange/TUCR_counts_80.txt

#Counts
multiBamCov -bams tcga_analysis/reindex/*bam -bed hg38.ultraConserved.bed -q 10 -f 0.80 >> tcga_analysis/foldchange/TUCR_counts_80.txt

#Write header line
echo -e chrom"\t"start"\t"end"\t"id *bam | sed -e "s/ /\t/g" > gene_counts.txt

#Counts
multiBamCov -bams *bam -bed CHESSgenes.bed -q 10 >> gene_counts.txt
```

## Acquiring total reads from RNA-Seq BAM files

```{bash, eval = FALSE}

mkdir tcga_analysis/readcounts/

module load samtools/1.9

for bam in $(find tcga_analysis/reindex/*.bam -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools view -c -F 260 $bam > $name.readcounts.txt
mv $name.readcounts.txt ./tcga_analysis/readcounts/
done

echo "id,counts" > tcga_analysis/readcounts/seqdepth_counts.csv 

for txt in $(find tcga_analysis/readcounts/* -name '*.txt')
do
name=$(echo $txt | awk -F ".readcounts.txt" '{print $1}')
echo $name
reads=$(head $txt)
echo "$name,$reads" >> tcga_analysis/readcounts/seqdepth_counts.csv 
done

```

## Acquiring clinical survival data

```{bash, eval = FALSE}

wget http://gdac.broadinstitute.org/runs/stddata__2016_01_28/data/GBM/20160128/gdac.broadinstitute.org_GBM.Merge_Clinical.Level_1.2016012800.0.0.tar.gz

tar -xvzf gdac.broadinstitute.org_GBM.Merge_Clinical.Level_1.2016012800.0.0.tar.gz

mv gdac.broadinstitute.org_GBM.Merge_Clinical.Level_1.2016012800.0.0 GBM.Merge_Clinical

```

## Acquiring SNPs

```{bash, eval = FALSE}
mkdir SNPs

cd SNPs

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_1.bed.gz
gunzip bed_chr_1.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_10.bed.gz
gunzip bed_chr_10.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_11.bed.gz
gunzip bed_chr_11.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_13.bed.gz
gunzip bed_chr_13.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_12.bed.gz
gunzip bed_chr_12.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_14.bed.gz
gunzip bed_chr_14.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_15.bed.gz
gunzip bed_chr_15.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_16.bed.gz
gunzip bed_chr_16.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_17.bed.gz
gunzip bed_chr_17.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_18.bed.gz
gunzip bed_chr_18.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_19.bed.gz
gunzip bed_chr_19.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_2.bed.gz
gunzip bed_chr_2.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_20.bed.gz
gunzip bed_chr_20.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_21.bed.gz
gunzip bed_chr_21.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_22.bed.gz
gunzip bed_chr_22.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_3.bed.gz
gunzip bed_chr_3.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_4.bed.gz
gunzip bed_chr_4.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_5.bed.gz
gunzip bed_chr_5.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_6.bed.gz
gunzip bed_chr_6.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_7.bed.gz
gunzip bed_chr_7.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_8.bed.gz
gunzip bed_chr_8.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_9.bed.gz
gunzip bed_chr_9.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_X.bed.gz
gunzip bed_chr_X.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_Y.bed.gz
gunzip bed_chr_Y.bed.gz

module load gcc/7.1.0
module load bedtools/2.26.0

for bed in *.bed
do
echo $bed
intersectBed -a hg38.ultraConserved.bed -b $bed -wa -wb > $bed.txt
done

cat *.txt > merged.txt
```

## Generating PsychoTUCRs

```{bash, eval = FALSE}

bedtools random -g hg38.genome -n 481 -l 243 -seed 71346

```

## Identifying differentially expressed TUCRs with DESeq2

```{r DESeq2_TUCRs, results="hide", message = FALSE}

tucrcounts <- read.table(countfilename,header = TRUE)
metadata <- read_csv(file = "tcga_metadata.csv")

tucrcounts2 <- tucrcounts[,5:ncol(tucrcounts)]

rownames(tucrcounts2) <- tucrcounts[,4]
tucrcounts.info <- tucrcounts[,1:4]
tucrcounts.info$length <- (tucrcounts$end - tucrcounts$start)/1000
tucrcounts <- cbind(tucrcounts.info,tucrcounts2)


dds_TUCR <- DESeqDataSetFromMatrix(countData = tucrcounts2,
                              colData = metadata,
                              design = ~dex)
dds_TUCR <- DESeq(dds_TUCR)
res_TUCR <- results(dds_TUCR, tidy=TRUE)
res_TUCR <- as_tibble(res_TUCR)

write.csv(res_TUCR, file = paste(outputdir,"/results_TUCR.csv",sep=""))

```

## Completing survival analysis for TUCRs

```{r surcor_TUCRs, message = FALSE}
if(!file.exists("GBM.clin.merged.txt")){unzip("GBM.clin.merged.zip")}

countdata <- read.table(countfilename,header = TRUE)
metadata <- read_csv(file = "tcga_metadata.csv")

posdata <- countdata[,1:4]
countdata <- countdata[,5:length(colnames(countdata))]

countdata <- as.matrix(countdata)

n_index <- c(seq(36,39),73)
t_index <- c(seq(1,35),seq(40,72),seq(74,ncol(countdata)))

vm <- function(x){
  cond <- factor(ifelse(seq(1,dim(x)[2],1) %in% t_index, 1,  0))
  d <- model.matrix(~1+cond)
  x <- t(apply(x,1,as.numeric))
  ex <- voom(x,d,plot=F)
  return(ex$E)
}

count_vm <- vm(countdata)
colnames(count_vm) <- metadata$barcode

scal <- function(x,y){
  mean_n <- rowMeans(y)  # mean of normal
  sd_n <- apply(y,1,sd)  # SD of normal
  # z score as (value - mean normal)/SD normal
  res <- matrix(nrow=nrow(x), ncol=ncol(x))
  colnames(res) <- colnames(x)
  rownames(res) <- rownames(x)
  for(i in 1:dim(x)[1]){
    for(j in 1:dim(x)[2]){
      res[i,j] <- (x[i,j]-mean_n[i])/sd_n[i]
    }
  }
  return(res)
}

z_rna <- scal(count_vm[,t_index],count_vm[,n_index])
rownames(z_rna) <- posdata[,4]

clinical <- read.table("GBM.clin.merged.txt",header = TRUE)
#clinical <- read.table("GBM.oncolnc.merged.txt",header = TRUE)

clinical$time <- as.numeric(clinical$time)

sum(clinical$patient %in% colnames(z_rna))

ind_tum <- which(unique(colnames(z_rna)) %in% clinical$patient)
ind_clin <- which(clinical$patient %in% colnames(z_rna))

out.tab <- c()
for(x in 1:nrow(count_vm)){
  ind_gene <- x
  s <- Surv(clinical$time[ind_clin],clinical$status[ind_clin])
  cx <- coxph(formula = s ~ count_vm[ind_gene,ind_tum])
  cx <- tidy(cx)
  out.tab <- rbind(out.tab,cx)
}

surv_TUCR <- cbind(posdata,out.tab)
surv_TUCR <- surv_TUCR %>% select(-term)

write_csv(surv_TUCR,paste(outputdir,"/survival_TUCRs.csv",sep=""))

if(makekpmplots == TRUE){

if(!dir.exists(paste(outputdir,"/tucrKPMplots",sep=""))){
  dir.create(paste(outputdir,"/tucrKPMplots",sep=""))
}

km_countdata <- countdata
colnames(km_countdata) <- metadata$barcode
posdata$median <- rowMedians(countdata)
km_TUCRs <- cbind(posdata,km_countdata)
km_TUCRs <- km_TUCRs[,4:ncol(km_TUCRs)]
km_TUCRs <- km_TUCRs %>%
  gather(key = "sample", value = "count",-id,-median) %>%
  mutate(group = ifelse(count>=median,"high","low")) %>%
  select(id,median,"patient"=sample,group) %>%
  left_join(clinical,by="patient") %>%
  filter(median != 0)

TUCRids <- posdata$id
i = 1

for(i in 1:length(TUCRids)){
  print(i)
  TUCR <- TUCRids[i]
  #TUCR <- "uc.110"
  kmdata <- km_TUCRs %>%
    filter(id == TUCR)
if(length(as.vector(kmdata$id))>0){
  fit <- survfit(Surv(time, status) ~ group, data = kmdata)
  p <- ggsurvplot(fit,data=kmdata,conf.int = TRUE,pval = TRUE,risk.table = TRUE)
  ggsave(file=paste(outputdir,"/tucrKPMplots/",TUCR,"_kpmplot.png",sep = ""),plot = print(p))}
}
}
```

## Generating volcano plot to visualize DE genes

```{r DESeq2plots_TUCRs, message = FALSE}

### Volcano plot

volcanoplot <- function (res,genes = "all",title = "Deregulated Genes", output = "results_volcanoplot.png",repel=TRUE,height = 5, width = 5, dpi = 300){
  #res <- res_TUCR
  #genes = c("uc.110","uc.62")
  #title = "Deregulated Genes"
  vres <- res
  vres <- vres %>% mutate(filter1 = abs(log2FoldChange)>=1,filter2 = padj <=0.05,filter3="",gene="")
  for (i in 1:length(vres$row)){
    if(is.na(vres$padj[i])){vres$filter3[i] <- "black"}else{
    if(vres$filter1[i] == TRUE && vres$filter2[i] == TRUE){
    vres$filter3[i] <- "red"
  }else{
      if(vres$filter1[i] == TRUE){
    vres$filter3[i] <- "blue"
      }else{vres$filter3[i] <- "black"}
  }}
    ifelse(genes!="all",ifelse(!is.na(match(vres$row[i],genes)),vres$gene[i] <- vres$row[i], ""),vres$gene[i] <- vres$row[i])
    }
  
  
  
  ## plot
if(repel==TRUE){
  p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=filter3)) +
        scale_color_manual(values = c("black","blue","red")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25))) +
        geom_text_repel(aes(x=log2FoldChange, y=-log10(padj),label = gene),force=50)
  }else{
   p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=filter3)) +
        scale_color_manual(values = c("black","blue","red")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25))) +
        geom_text(aes(x=log2FoldChange, y=-log10(padj),label = gene), label.size = 0.15)
  }
ggsave(output, width = width, height = height, dpi = dpi)
}

volcanoplot(res_TUCR,genes=c("uc.62","uc.110","uc.193","uc.210","uc.427","uc.212","uc.341"),title = "TUCR deregulation in GBM", output = paste(outputdir,"/myron_tucr_results_volcanoplot.png",sep=""))

volcanoplot(res_TUCR,title = "TUCR deregulation in GBM", output = paste(outputdir,"/tucr_results_volcanoplot.png",sep=""),repel = FALSE,width=7.5)

```

## Generate RPKMs for filtering

```{r RPKM_TUCRs, results="hide"}

rpkmcounts <- tucrcounts[,6:ncol(tucrcounts)]
rpkmcounts.info <- tucrcounts[,1:5]
rownames(rpkmcounts) <- tucrcounts$id

genelength <- tucrcounts$length

seqdepth <- read.csv(file = "seqdepth_counts.csv",row.names=1)

seqdepth <- as.vector(seqdepth$counts)

rpkm <- function(counts,len,dep){
  x <- counts/len
  return(t(t(x)/dep))
}

rpkm.TUCRs <- rpkm(rpkmcounts,genelength,seqdepth)
rpkm.TUCRs2 <- cbind(rpkmcounts.info,rpkm.TUCRs)

write.csv(rpkm.TUCRs2,paste(outputdir,"/rpkm.TUCRs.csv",sep=""))
```

## Generate RPKM Box Plots for each TUCR

```{r boxplot_TUCRs, results = "hide"}
if(makeRPKMboxplots == TRUE){
  
if(!dir.exists(paste(outputdir,"/RPKMboxplots",sep=""))){
  dir.create(paste(outputdir,"/RPKMboxplots",sep=""))
}
rpkm.dotplot <- rpkm.TUCRs2[,6:ncol(rpkm.TUCRs2)]
TUCRids <- rpkm.TUCRs2[,4]
rpkm.dotplot <- cbind(TUCRids,rpkm.dotplot)

i = 1
for(i in 1:length(TUCRids)){
#TUCR <- "uc.110"
TUCR <- TUCRids[i]
print(i)
rpkm.dotplot2 <- rpkm.dotplot %>% 
  gather(key = "Sample", value = "RPKM",-TUCRids) %>%
  filter(TUCRids == TUCR)

p<- ggplot(rpkm.dotplot2, aes(x=TUCRids, y=RPKM)) +
  geom_boxplot() +
  geom_dotplot(binaxis='y', stackdir='center',
               stackratio=0.75, dotsize=0.5) + 
  ggtitle(paste("RPKM expression of ",TUCR)) +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(y="Reads per kilobase million (RPKM)", x = "TUCR") +
  stat_summary(fun.y=mean, geom="point", shape=20, size=10, color="red", fill="red")

ggsave(p,file=paste(outputdir,"/RPKMboxplots/",TUCR,"_rpkm_boxplot.png",sep = ""))
}

}
```

## Generate heatmap for all TUCRs

```{r heatmap_TUCRs}
heatmap.TUCRs <- rpkm.TUCRs2[,6:ncol(rpkm.TUCRs2)]
TUCRids <- rpkm.TUCRs2[,4]
means <- apply(heatmap.TUCRs,1,mean)

heatmap.TUCRs2 <- as.data.frame(apply(heatmap.TUCRs,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.TUCRs2 <- cbind(TUCRids,means,heatmap.TUCRs2)

heatmap.TUCRs2 <- heatmap.TUCRs2 %>% 
  gather(key = "Sample", value = "RPKM",-TUCRids,-means)

p <- ggplot(data = heatmap.TUCRs2,mapping = aes(x = Sample,y = reorder(TUCRids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/TUCR_heatmap.png",sep=""))
```

## Generate heatmap for my TUCRs of interest

```{r heatmap_Myron}
heatmap.TUCRs <- rpkm.TUCRs2[,6:ncol(rpkm.TUCRs2)]
TUCRids <- rpkm.TUCRs2[,4]
means <- apply(heatmap.TUCRs,1,mean)

heatmap.TUCRs2 <- as.data.frame(apply(heatmap.TUCRs,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.TUCRs2 <- cbind(TUCRids,means,heatmap.TUCRs2)

heatmap.TUCRs2 <- heatmap.TUCRs2 %>% 
  gather(key = "Sample", value = "RPKM",-TUCRids,-means) %>% 
  filter(TUCRids == "uc.62"|TUCRids == "uc.110"|TUCRids == "uc.193"|TUCRids == "uc.210"|TUCRids == "uc.427"|TUCRids == "uc.212"|TUCRids == "uc.341")

p <- ggplot(data = heatmap.TUCRs2,mapping = aes(x = Sample,y = reorder(TUCRids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/Myron_heatmap.png",sep=""))
```

## Identifying differentially expressed stringtie transcripts containing TUCRs with DESeq2

```{r DESeq2_stringtie, results="hide", message = FALSE}

stringtiecounts <- read.table("stringtie_TUCR_counts.txt",header = TRUE)
stringtiecounts <- distinct(stringtiecounts)
metadata <- read_csv(file = "tcga_metadata.csv")

stringtiecounts2 <- stringtiecounts[,5:ncol(stringtiecounts)]
rownames(stringtiecounts2) <- stringtiecounts[,4]
stringtiecounts.info <- stringtiecounts[,1:4]
stringtiecounts.info$length <- (stringtiecounts$end - stringtiecounts$start)/1000
stringtiecounts <- cbind(stringtiecounts.info,stringtiecounts2)

dds_stringtie <- DESeqDataSetFromMatrix(countData = stringtiecounts2,
                              colData = metadata,
                              design = ~dex)
dds_stringtie <- DESeq(dds_stringtie)
res_stringtie <- results(dds_stringtie, tidy=TRUE)
res_stringtie <- as_tibble(res_stringtie)

write.csv(res_stringtie, file = paste(outputdir,"/results_stringtie.csv",sep=""))

```

## Completing survival analysis for stringtie transcripts containing TUCRs

```{r surcor_stringtie, message = FALSE}
if(!file.exists("GBM.clin.merged.txt")){unzip("GBM.clin.merged.zip")}

countdata <- read.table("stringtie_TUCR_counts.txt",header = TRUE)
metadata <- read_csv(file = "tcga_metadata.csv")

posdata <- countdata[,1:4]
countdata <- countdata[,5:length(colnames(countdata))]

countdata <- as.matrix(countdata)


n_index <- c(seq(36,39),73)
t_index <- c(seq(1,35),seq(40,72),seq(74,ncol(countdata)))

vm <- function(x){
  cond <- factor(ifelse(seq(1,dim(x)[2],1) %in% t_index, 1,  0))
  d <- model.matrix(~1+cond)
  x <- t(apply(x,1,as.numeric))
  ex <- voom(x,d,plot=F)
  return(ex$E)
}

count_vm <- vm(countdata)
colnames(count_vm) <- metadata$barcode


scal <- function(x,y){
  mean_n <- rowMeans(y)  # mean of normal
  sd_n <- apply(y,1,sd)  # SD of normal
  # z score as (value - mean normal)/SD normal
  res <- matrix(nrow=nrow(x), ncol=ncol(x))
  colnames(res) <- colnames(x)
  rownames(res) <- rownames(x)
  for(i in 1:dim(x)[1]){
    for(j in 1:dim(x)[2]){
      res[i,j] <- (x[i,j]-mean_n[i])/sd_n[i]
    }
  }
  return(res)
}

z_rna <- scal(count_vm[,t_index],count_vm[,n_index])
rownames(z_rna) <- posdata[,4]

clinical <- read.table("GBM.clin.merged.txt",header = TRUE)
clinical$time <- as.numeric(clinical$time)

sum(clinical$patient %in% colnames(z_rna))

ind_tum <- which(unique(colnames(z_rna)) %in% clinical$patient)
ind_clin <- which(clinical$patient %in% colnames(z_rna))

out.tab <- c()
for(x in 1:nrow(count_vm)){
  ind_gene <- x
  s <- Surv(clinical$time[ind_clin],clinical$status[ind_clin])
  cx <- coxph(formula = s ~ count_vm[ind_gene,ind_tum])
  cx <- tidy(cx)
  out.tab <- rbind(out.tab,cx)
}

surv_stringtie <- cbind(posdata,out.tab)
surv_stringtie <- surv_stringtie %>% select(-term)

write_csv(surv_stringtie,paste(outputdir,"/survival_TUCRs_stringtie.csv",sep=""))
```

## Generating volcano plot to visualize DE stringtie transcripts

```{r DESeq2plots_stringtie, message = FALSE}

### Volcano plot

volcanoplot <- function (res,title = "Deregulated Genes", output = "results_volcanoplot.png"){
  #res <- res_TUCR
  #genes = c("uc.110","uc.62")
  #title = "Deregulated Genes"
  vres <- res
  vres <- vres %>% mutate(filter1 = abs(log2FoldChange)>=1,filter2 = padj <=0.05,filter3="")#,gene="")
  for (i in 1:length(vres$row)){
    if(is.na(vres$padj[i])){vres$filter3[i] <- "black"}else{
    if(vres$filter1[i] == TRUE && vres$filter2[i] == TRUE){
    vres$filter3[i] <- "red"
  }else{
      if(vres$filter1[i] == TRUE){
    vres$filter3[i] <- "blue"
      }else{vres$filter3[i] <- "black"}
  }}
    #ifelse(genes!="all",ifelse(!is.na(match(vres$row[i],genes)),vres$gene[i] <- vres$row[i], ""),vres$gene[i] <- vres$row[i])
    }
  
  
  
  ## plot
  p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=filter3)) +
        scale_color_manual(values = c("black","blue","red")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25)))
  
  ggsave(output)
}

volcanoplot(res_stringtie,title = "Deregulation of stringtie transcripts containing TUCRs in GBM", output = paste(outputdir,"/stringtie_results_volcanoplot.png",sep=""))

```

## Generate stringtie RPKMs for filtering

```{r RPKM_stringtie, results="hide"}

stringtie.rpkmcounts <- stringtiecounts[,6:ncol(stringtiecounts)]
stringtie.rpkmcounts.info <- stringtiecounts[,1:5]
rownames(stringtie.rpkmcounts) <- stringtiecounts$id

genelength <- stringtiecounts$length

seqdepth <- read.csv(file = "seqdepth_counts.csv",row.names=1)

seqdepth <- as.vector(seqdepth$counts)

rpkm <- function(counts,len,dep){
  x <- counts/len
  return(t(t(x)/dep))
}

rpkm.stringtie <- rpkm(stringtie.rpkmcounts,genelength,seqdepth)
rpkm.stringtie2 <- cbind(stringtie.rpkmcounts.info,rpkm.stringtie)

write.csv(rpkm.stringtie2,paste(outputdir,"/rpkm.stringtie.csv",sep=""))
```

## Generate heatmap for all stringtie TUCR transcripts

```{r heatmap_stringtie, message=FALSE}
heatmap.stringtie <- rpkm.stringtie2[,6:ncol(rpkm.stringtie2)]
stringtieids <- rpkm.stringtie2[,5]
means <- apply(heatmap.stringtie,1,mean)

heatmap.stringtie2 <- as.data.frame(apply(heatmap.stringtie,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.stringtie2 <- cbind(stringtieids,means,heatmap.stringtie2)

heatmap.stringtie2 <- heatmap.stringtie2 %>% 
  gather(key = "Sample", value = "RPKM",-stringtieids,-means)

p <- ggplot(data = heatmap.stringtie2,mapping = aes(x = Sample,y = reorder(stringtieids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/stringtie_heatmap.png",sep=""))
```

## Identifying differentially expressed CHESS genes containing TUCRs with DESeq2

```{r DESeq2_CHESS, results="hide", message = FALSE}

CHESScounts <- read.table("intersectchess_TUCR_counts.txt",header = TRUE)
CHESScounts <- distinct(CHESScounts)
metadata <- read_csv(file = "tcga_metadata.csv")

CHESScounts2 <- CHESScounts[,5:ncol(CHESScounts)]
rownames(CHESScounts2) <- CHESScounts[,4]
CHESScounts.info <- CHESScounts[,1:4]
CHESScounts.info$length <- (CHESScounts$end - CHESScounts$start)/1000
CHESScounts <- cbind(CHESScounts.info,CHESScounts2)

dds_CHESS <- DESeqDataSetFromMatrix(countData = CHESScounts2,
                              colData = metadata,
                              design = ~dex)
dds_CHESS <- DESeq(dds_CHESS)
res_CHESS <- results(dds_CHESS, tidy=TRUE)
res_CHESS <- as_tibble(res_CHESS)

write.csv(res_CHESS, file = paste(outputdir,"/results_CHESS.csv",sep=""))

```

## Generating volcano plot to visualize DE CHESS genes containing TUCRs

```{r DESeq2plots_CHESS, message = FALSE}

### Volcano plot

volcanoplot <- function (res,title = "Deregulated Genes", output = "results_volcanoplot.png"){
  #res <- res_TUCR
  #genes = c("uc.110","uc.62")
  #title = "Deregulated Genes"
  vres <- res
  vres <- vres %>% mutate(filter1 = abs(log2FoldChange)>=1,filter2 = padj <=0.05,filter3="")#,gene="")
  for (i in 1:length(vres$row)){
    if(is.na(vres$padj[i])){vres$filter3[i] <- "black"}else{
    if(vres$filter1[i] == TRUE && vres$filter2[i] == TRUE){
    vres$filter3[i] <- "red"
  }else{
      if(vres$filter1[i] == TRUE){
    vres$filter3[i] <- "blue"
      }else{vres$filter3[i] <- "black"}
  }}
    #ifelse(genes!="all",ifelse(!is.na(match(vres$row[i],genes)),vres$gene[i] <- vres$row[i], ""),vres$gene[i] <- vres$row[i])
    }
  
  
  
  ## plot
  p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=filter3)) +
        scale_color_manual(values = c("black","blue","red")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25)))
  
  ggsave(output)
}

volcanoplot(res_CHESS,title = "Deregulation of CHESS genes containing TUCRs in GBM", output = paste(outputdir,"/CHESS_results_volcanoplot.png",sep=""))

```

## Completing survival analysis for stringtie transcripts containing TUCRs

```{r surcor_CHESS, message = FALSE}
if(!file.exists("GBM.clin.merged.txt")){unzip("GBM.clin.merged.zip")}

countdata <- read.table("intersectchess_TUCR_counts.txt",header = TRUE)
metadata <- read_csv(file = "tcga_metadata.csv")

posdata <- countdata[,1:4]
countdata <- countdata[,5:length(colnames(countdata))]

countdata <- as.matrix(countdata)

n_index <- c(seq(36,39),73)
t_index <- c(seq(1,35),seq(40,72),seq(74,ncol(countdata)))

vm <- function(x){
  cond <- factor(ifelse(seq(1,dim(x)[2],1) %in% t_index, 1,  0))
  d <- model.matrix(~1+cond)
  x <- t(apply(x,1,as.numeric))
  ex <- voom(x,d,plot=F)
  return(ex$E)
}

count_vm <- vm(countdata)
colnames(count_vm) <- metadata$barcode


scal <- function(x,y){
  mean_n <- rowMeans(y)  # mean of normal
  sd_n <- apply(y,1,sd)  # SD of normal
  # z score as (value - mean normal)/SD normal
  res <- matrix(nrow=nrow(x), ncol=ncol(x))
  colnames(res) <- colnames(x)
  rownames(res) <- rownames(x)
  for(i in 1:dim(x)[1]){
    for(j in 1:dim(x)[2]){
      res[i,j] <- (x[i,j]-mean_n[i])/sd_n[i]
    }
  }
  return(res)
}

z_rna <- scal(count_vm[,t_index],count_vm[,n_index])
rownames(z_rna) <- posdata[,4]

clinical <- read.table("GBM.clin.merged.txt",header = TRUE)
clinical$time <- as.numeric(clinical$time)

sum(clinical$patient %in% colnames(z_rna))

ind_tum <- which(unique(colnames(z_rna)) %in% clinical$patient)
ind_clin <- which(clinical$patient %in% colnames(z_rna))

out.tab <- c()
for(x in 1:nrow(count_vm)){
  ind_gene <- x
  s <- Surv(clinical$time[ind_clin],clinical$status[ind_clin])
  cx <- coxph(formula = s ~ count_vm[ind_gene,ind_tum])
  cx <- tidy(cx)
  out.tab <- rbind(out.tab,cx)
}

surv_CHESS <- cbind(posdata,out.tab)
surv_CHESS <- surv_CHESS %>% select(-term)

write_csv(surv_CHESS,paste(outputdir,"/survival_TUCRs_CHESS.csv",sep=""))

```

## Generate CHESS gene RPKMs for filtering

```{r RPKM_CHESS, results="hide"}

CHESS.rpkmcounts <- CHESScounts[,6:ncol(CHESScounts)]
CHESS.rpkmcounts.info <- CHESScounts[,1:5]
rownames(CHESS.rpkmcounts) <- CHESScounts$id

genelength <- CHESScounts$length

seqdepth <- read.csv(file = "seqdepth_counts.csv",row.names=1)

seqdepth <- as.vector(seqdepth$counts)

rpkm <- function(counts,len,dep){
  x <- counts/len
  return(t(t(x)/dep))
}

rpkm.CHESS <- rpkm(CHESS.rpkmcounts,genelength,seqdepth)
rpkm.CHESS2 <- cbind(CHESS.rpkmcounts.info,rpkm.CHESS)

write.csv(rpkm.CHESS2,paste(outputdir,"/rpkm.CHESS.csv",sep=""))
```

## Generate heatmap for all CHESS genes containing TUCRs

```{r heatmap_CHESS, message=FALSE}
heatmap.CHESS <- rpkm.CHESS2[,6:ncol(rpkm.CHESS2)]
CHESSids <- rpkm.CHESS2[,5]
means <- apply(heatmap.CHESS,1,mean)

heatmap.CHESS2 <- as.data.frame(apply(heatmap.CHESS,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.CHESS2 <- cbind(CHESSids,means,heatmap.CHESS2)

heatmap.CHESS2 <- heatmap.CHESS2 %>% 
  gather(key = "Sample", value = "RPKM",-CHESSids,-means)

p <- ggplot(data = heatmap.CHESS2,mapping = aes(x = Sample,y = reorder(CHESSids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/CHESS_heatmap.png",sep=""))
```

## Visualize Stephen and Alex's FPKMs to compare to my own RPKMs

```{r sdt_heatmap, message=FALSE}
fpkm.sdt <- read_csv("aboundaer_tucr_gbm-survival_fpkms.csv")
heatmap.sdt <- fpkm.sdt[,2:ncol(fpkm.sdt)]
sdtids <- fpkm.sdt[,1]
means <- apply(heatmap.sdt,1,mean)

heatmap.sdt2 <- as.data.frame(apply(heatmap.sdt,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.sdt2 <- cbind(sdtids,means,heatmap.sdt2)

heatmap.sdt2 <- heatmap.sdt2 %>% 
  gather(key = "Sample", value = "FPKM",-id,-means)

p <- ggplot(data = heatmap.sdt2,mapping = aes(x = Sample,y = reorder(id,means), fill = FPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/sdt_heatmap.png",sep=""))
```

## Generate figures for cell fractionation

```{r cellfrac_TUCR, eval=FALSE,message=FALSE}

if(!dir.exists(paste(outputdir,"/cellfrac",sep=""))){
  dir.create(paste(outputdir,"/cellfrac",sep=""))
}

CellFrackin <- function(filename){
all_content = readLines(filename)
skip_second = all_content[-c(1:19)]
x <- read.csv(textConnection(skip_second), header = TRUE, stringsAsFactors = FALSE)
samplenames <- t(as.data.frame(strsplit(as.character(x$Sample), " ")))
colnames(samplenames) <- c("Line","Fraction")
x <- cbind(x,samplenames)
x<- x %>%
    select(Well,Target,Sample,Line,Fraction,Cq) %>%
    group_by(Target,Line,Fraction) %>%
    summarize(meanCq = mean(Cq)) %>%
    mutate(dct = meanCq - min(meanCq)) %>%
    mutate(FoldChange = 2^(-dct)) %>%
    mutate(ID = paste(Target,Line,Fraction,sep=" "))

p <- ggplot(x,aes(x = Target, y = FoldChange, fill = Fraction)) + 
  labs(title = "Cellular localization of TUCRs in GBM cell lines") +
  ylab("Fold Change") +
  geom_bar(stat = "identity",position="dodge") + 
  theme(axis.text.x = element_text(angle = -90, size = 10),
        axis.text.y = element_text(size = 10),
        axis.ticks.x = element_blank(),
        plot.title = element_text(size = 15, face = "bold",hjust = 0.5),
        axis.title.y = element_text("Fold Change",size = 10)) +
  facet_wrap(~Line)

ggsave(p,file=paste(outputdir,"/",filename,".png",sep=""), width = 7.5, height = 5, dpi = 300)
x %>%
  select(-ID) %>%
  write_csv(paste(outputdir,"/",filename," (Output).csv",sep=""))
}

path <- list.files("cellfrac")
i = 1

for(i in 1:length(path)){
  print(i)
  CellFrackin(paste("cellfrac",path[i],sep="/"))
}

```

## Identify DE TUCRs that are correlated with each other

```{r correlations, eval=FALSE}

if(!dir.exists(paste(outputdir,"/correlations",sep=""))){
  dir.create(paste(outputdir,"/correlations",sep=""))
}

if(useintermediate==FALSE){
TUCRids <- read.table(countfilename,header = TRUE)
TUCRids <- as.character(TUCRids[,4])
tucrcounts <- read.table("mergedcounts.txt",header = TRUE)

metadata <- read_csv(file = "tcga_metadata.csv")



tucr.cor <- tucrcounts[,5:ncol(tucrcounts)]
rownames(tucr.cor) <- as.character(tucrcounts[,4])

posdata <- tucrcounts[,1:4]

countdata <- as.matrix(tucr.cor)

n_index <- c(seq(36,39),73)
t_index <- c(seq(1,35),seq(40,72),seq(74,ncol(countdata)))

vm <- function(x){
  cond <- factor(ifelse(seq(1,dim(x)[2],1) %in% t_index, 1,  0))
  d <- model.matrix(~1+cond)
  x <- t(apply(x,1,as.numeric))
  ex <- voom(x,d,plot=F)
  return(ex$E)
}

count_vm <- vm(countdata)
colnames(count_vm) <- metadata$barcode
posdata$median <- rowMedians(as.matrix(count_vm)) 

tucr.cor <- cbind(posdata[,4:5],count_vm)

tucr.cor.matrix <- data.frame(row.names = posdata[,4])


i=3

for(i in 3:ncol(tucr.cor)){
  print(i)
  tucr.cor.matrix[,i-2] <- tucr.cor[,i]/tucr.cor[,2]
}

colnames(tucr.cor.matrix) <- metadata$id
tucr.cor.matrix2 <- t(as.data.frame(apply(as.data.frame(tucr.cor.matrix),1:2,function(x) {ifelse(x>=1,x,-(1/x))})))
tucr.cor.matrix2 <- as.data.frame(tucr.cor.matrix2)


tucr.cor.matrix3 <- cor(tucr.cor.matrix2,method="spearman")

geneindex <- rownames(tucr.cor.matrix)
save(tucr.cor.matrix3,geneindex,TUCRids,file="matrixintermediate.RData")
rm(list=ls())
load(file="matrixintermediate.RData")}else{
  load(file="matrixintermediate.RData")
}

summarycor <- data.frame(searchTUCR=character(0),ngenes=integer(0),meancor=numeric(0),mediancor=numeric(0),mincor=numeric(0),maxcor=numeric(0))

i = 1

for(i in 1:length(TUCRids)){
  print(i)
  searchTUCR <- TUCRids[i]
  #searchTUCR <- "uc.110"
###uc.110 is at row 20972 in tucr.cor.matrix3 via manual confirmation.
tucr.cor.matrix.match <- tucr.cor.matrix3[match(searchTUCR,geneindex),]
tucr.cor.matrix.match <- as.data.frame(tucr.cor.matrix.match)
tucr.cor.matrix.match <- tucr.cor.matrix.match %>% 
  mutate(partner = geneindex) %>%
  mutate(gene = searchTUCR) %>%
  select(gene,partner,"cor"=tucr.cor.matrix.match) %>%
  filter(abs(cor)>=0.3 & partner != searchTUCR) %>%
  readr::write_excel_csv(here::here(paste(outputdir,"/correlations",sep=""),paste(searchTUCR,"_correlations.csv",sep="")))
meancor <- mean(tucr.cor.matrix.match$cor)
mediancor <- median(tucr.cor.matrix.match$cor)
ngenes <- length(tucr.cor.matrix.match$cor)
mincor <- min(tucr.cor.matrix.match$cor)
maxcor <- max(tucr.cor.matrix.match$cor)
newrow <- cbind(searchTUCR,ngenes,meancor,mediancor,mincor,maxcor)
summarycor <- rbind(summarycor,newrow)}

summarycor %>%
readr::write_excel_csv(here::here(paste(outputdir,"/correlations",sep=""),paste("cor_summary.csv",sep="")))

```

## Make figures for fold change

```{r overexp_TUCR, eval = FALSE, message=FALSE}

overexp_TUCR_cp <- c("#C77CFF","#F8766D","#619CFF","#00BA38","#FFFF00")

filename <- "tucroverexpression_U87.csv"
x <- read.csv(filename)

p <- ggplot(x,aes(x = Target, y = FC, fill = Sample)) + 
  labs(title = "TUCR Overexpression in U87 cells") +
  ylab("Fold Change") +
  geom_bar(stat = "identity",position="dodge",colour="black") + 
  theme(axis.text.x = element_text(angle = -90, size = 10),
        axis.text.y = element_text(size = 10),
        axis.ticks.x = element_blank(),
        plot.title = element_text(size = 15, face = "bold",hjust = 0.5),
        axis.title.y = element_text("Fold Change",size = 10)) +
  facet_zoom(ylim = c(0, 2)) +
  scale_fill_manual(values=overexp_TUCR_cp)

ggsave(p,file=paste(outputdir,"/",filename,".png",sep=""), width = 7.5, height = 5, dpi = 300)


```

## Volcano plot for intergenic TUCRs only

```{r DESeq2plots_intergenicTUCRs, message = FALSE}

### Volcano plot

res_intergenic <- read.csv("results_TUCR_intergenic_sdt.csv")

volcanoplot <- function (res,genes = "all",title = "Deregulated Genes", output = "results_volcanoplot.png",repel=TRUE,height = 5, width = 5, dpi = 300){
  #res <- res_TUCR
  #genes = c("uc.110","uc.62")
  #title = "Deregulated Genes"
  vres <- res
  vres <- vres %>% mutate(filter1 = abs(log2FoldChange)>=1,filter2 = padj <=0.05,filter3="",gene="")
  for (i in 1:length(vres$row)){
    if(is.na(vres$padj[i])){vres$filter3[i] <- "black"}else{
    if(vres$filter1[i] == TRUE && vres$filter2[i] == TRUE){
    vres$filter3[i] <- "red"
  }else{
      if(vres$filter1[i] == TRUE){
    vres$filter3[i] <- "blue"
      }else{vres$filter3[i] <- "black"}
  }}
    ifelse(genes!="all",ifelse(!is.na(match(vres$row[i],genes)),vres$gene[i] <- as.character(vres$row[i]), ""),vres$gene[i] <- as.character(vres$row[i]))
    }
  
  
  
  ## plot
if(repel==TRUE){
  p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=filter3)) +
        scale_color_manual(values = c("black","red","red")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25))) +
        geom_text_repel(aes(x=log2FoldChange, y=-log10(padj),label = gene),force=50)
  }else{
   p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=filter3)) +
        scale_color_manual(values = c("black","blue","red")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25))) +
        geom_text(aes(x=log2FoldChange, y=-log10(padj),label = gene), label.size = 0.15)
  }
ggsave(output, width = width, height = height, dpi = dpi)
}

volcanoplot(res_intergenic,genes=c("uc.62","uc.110","uc.193","uc.210","uc.427","uc.382","uc.403","uc.159"),title = "TUCR deregulation in GBM", output = paste(outputdir,"/intergenic_tucr_results_volcanoplot.png",sep=""))

```